### PR TITLE
Move lr tensor to cuda if needed

### LINKED ISSUE #119026: Adam(W) tests disabled for inductor/test_compiled_optimizers
This should be a temporary stopgap. I am trying to land my PR #118326 to correct many things, one of which is that the `optim_inputs_func`s have been incorrect and failing to generate capturable configs.

Fixing that and thus bringing forth the capturable configs leads to the following failure for 4 tests (after a land race), which I am going to disable in the PR (but we should re-enable):
```
======================================================================
FAIL: test_adam_amsgrad_capturable_cuda (__main__.CompiledOptimizerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2697, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 245, in test_fn
    self.check_cudagraphs_ran()
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 329, in check_cudagraphs_ran
    self.assertIsNotNone(manager)
AssertionError: unexpectedly None

To execute this test, run the following from the base repo dir:
     python test/inductor/test_compiled_optimizers.py -k test_adam_amsgrad_capturable_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
FAIL: test_adam_foreach_amsgrad_capturable_cuda (__main__.CompiledOptimizerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2697, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 245, in test_fn
    self.check_cudagraphs_ran()
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 329, in check_cudagraphs_ran
    self.assertIsNotNone(manager)
AssertionError: unexpectedly None

To execute this test, run the following from the base repo dir:
     python test/inductor/test_compiled_optimizers.py -k test_adam_foreach_amsgrad_capturable_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
FAIL: test_adamw_amsgrad_capturable_cuda (__main__.CompiledOptimizerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2697, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 245, in test_fn
    self.check_cudagraphs_ran()
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 329, in check_cudagraphs_ran
    self.assertIsNotNone(manager)
AssertionError: unexpectedly None

To execute this test, run the following from the base repo dir:
     python test/inductor/test_compiled_optimizers.py -k test_adamw_amsgrad_capturable_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

======================================================================
FAIL: test_adamw_foreach_amsgrad_capturable_cuda (__main__.CompiledOptimizerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py", line 2697, in wrapper
    method(*args, **kwargs)
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 245, in test_fn
    self.check_cudagraphs_ran()
  File "/data/users/janeyx/pytorch/test/inductor/test_compiled_optimizers.py", line 329, in check_cudagraphs_ran
    self.assertIsNotNone(manager)
AssertionError: unexpectedly None

To execute this test, run the following from the base repo dir:
     python test/inductor/test_compiled_optimizers.py -k test_adamw_foreach_amsgrad_capturable_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 212 tests in 496.567s

FAILED (failures=4)
```

I'd like to land the PR ASAP to make sure we're testing all the configs + to prevent further fallthrough.

cc @ezyang @gchanan @zou3519 @kadeng @mlazos

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/119026

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: test/inductor/test_compiled_optimizers.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import torch
 
 import torch._inductor
+import torch._inductor.cudagraph_trees
```
Comment (eellison) [inline test/inductor/test_compiled_optimizers.py:14]: not intended ?

**Review by eellison — APPROVED** [review]

File: test/inductor/test_compiled_optimizers.py (hunk 1 above)
Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:14]: this actually is intended, sometimes the import is missing?

Comment (janeyx99) [reply test/inductor/test_compiled_optimizers.py:14]: Looks like lines 13-15 are mergeable into one import

File: test/inductor/test_compiled_optimizers.py [hunk 2]
```diff
@@ -208,6 +200,9 @@ def test_fn(self):
             if run_cudagraphs:
                 stack.enter_context(config.patch({"triton.cudagraphs": True}))
 
+            if isinstance(kwargs.get("lr", None), torch.Tensor):
```
Comment (janeyx99) [inline test/inductor/test_compiled_optimizers.py:203]: Does this change suggest that the responsibility would be on the user to move their lr to the right device? What happens if there are mixed devices among the parameters of the optimizer?

Comment (mlazos) [reply test/inductor/test_compiled_optimizers.py:203]: Basically it just can't be CPU (which I'm not sure we test in the optimizer cudagraphs tests). We could possibly add something to the compiler cudagraphs impl to detect that they're CPU and 0-D but currently we basically print a warning that cudagraphs didn't run, and the line that causes it.

**Pull request closed without merging**